java/compaction/compaction-task-creation/src/main/java/sleeper/compaction/task/creation/RunTasks.java

```diff
@@ -110,60 +114,75 @@ public RunTasks(AmazonSQS sqsClient,
             requirements = Pair.of(requirements.getLeft(), (int) (requirements.getRight() * 0.95));
         }
 
-        this.scaler = new Scaler(asClient, ecsClient, autoScalingGroupName, this.clusterName,
-                requirements.getLeft(),
-                requirements.getRight());
+        return new EC2Scaler(asClient, ecsClient, instanceProperties.get(COMPACTION_AUTO_SCALING_GROUP),
+                instanceProperties.get(COMPACTION_CLUSTER), requirements.getLeft(), requirements.getRight());
+    }
+
+    public interface TaskCounts {
+        int getRunningAndPending(String clusterName);
+    }
+
+    public interface TaskLauncher {
+        void launchTasks(long startTime, int numberOfTasksToCreate);
     }
 
     public void run() {
         long startTime = System.currentTimeMillis();
         LOGGER.info("Queue URL is {}", sqsJobQueueUrl);
         // Find out number of messages in queue that are not being processed
-        int queueSize = QueueMessageCount.withSqsClient(sqsClient).getQueueMessageCount(sqsJobQueueUrl)
+        int queueSize = queueMessageCount.getQueueMessageCount(sqsJobQueueUrl)
                 .getApproximateNumberOfMessages();
         LOGGER.info("Queue size is {}", queueSize);
+        // Request 1 task for each item on the queue
+        run(startTime, queueSize);
+    }
+
+    public void run(int requestedTasks) {
+        run(System.currentTimeMillis(), requestedTasks);
+    }
 
+    private void run(long startTime, int requestedTasks) {
+        if (requestedTasks == 0) {
+            LOGGER.info("Finishing as number of tasks requested was 0");
+            return;
+        }
         // Find out number of pending and running tasks
-        int numRunningAndPendingTasks = CommonJobUtils.getNumPendingAndRunningTasks(clusterName, ecsClient);
+        int numRunningAndPendingTasks = taskCounts.getRunningAndPending(clusterName);
         LOGGER.info("Number of running and pending tasks is {}", numRunningAndPendingTasks);
 
+        if (numRunningAndPendingTasks >= maximumRunningTasks) {
+            LOGGER.info("Finishing as maximum running tasks of {} has been reached", maximumRunningTasks);
+            return;
+        }
         int maxNumTasksToCreate = maximumRunningTasks - numRunningAndPendingTasks;
-        LOGGER.info("Maximum number of tasks to create is {}", maxNumTasksToCreate);
-
-        // Do we need to scale?
+        LOGGER.info("Maximum concurrent tasks is {}", maxNumTasksToCreate);
+        LOGGER.info("Maximum number of tasks that can be created is {}", maxNumTasksToCreate);
```

Comment:
This is logging the same number twice. Should the first one be maximumRunningTasks?